xmodule/docs/decisions/0004-evaluate-wasm-safe-exec.rst

```diff
@@ -0,0 +1,107 @@
+4. Adopt WebAssembly as an Option for Safe Code Execution
+#########################################################
+
+Status
+******
+
+**Provisional**
+
+Subject to change due to implementation learnings and stakeholder feedback.
+
+Context
+*******
+
+The Open edX platform allows advanced course authors to write customized problem scoring code in Python. This has long been an area of strength, but it comes with significant operational issues, security risk, and long term maintenance burdens for both the authors and platform developers.
+
+Recent advancements in WebAssembly (Wasm) give us an opportunity to give course teams even more powerful capabilities for trusted server-side grading, while being safer, faster, and easier to maintain. Third party package support is not yet at a level where we can make a drop-in replacement for codejail, but we can still build the infrastructure that would make it possible, as well as offer exciting new capabilities.
+
+Why WebAssembly?
+================
+
+Code sandboxing currently happens using codejail, an AppArmor-based sandboxing mechanism that either runs on the same server as Studio/LMS, or runs on a separate server. This current system has worked for us for some time, but it has a number of drawbacks that WebAssembly is well positioned to address.
+
+Security
+--------
+
+In the current system, simple misconfigurations can be dangerous, since we're running a full instance of Python and relying on properly setting up codejail and AppArmor to constrain it.
+
+WebAssembly was created with safe execution in mind. When running Wasm code, it will by default have no access to the network or host filesystem.
+
+Maintenance
+-----------
+
+Content lasts longer than software, and authors don't expect to have to update their scoring code. When Python 2.7 was approaching its end-of-life, it was a large effort to get people to test their content with Python 3. Some of that content is probably forever broken. While this is an extreme example, there will be smaller breakages over time–libraries that get upgraded because the old versions are no longer available, Python versions that get updated because operating systems stop offering end-of-life versions, etc.
```

Comment:
Keeping this behind a Waffle flag sounds great.

What could a rough estimated timeline for discontinuing the codejail if these experiments go smoothly with WASM. What provoked this thought was that during the transition from codejail to wasm we will have to:

1. Build and generate binaries to support existing custom scoring problems and update them to use the new attributes so that they can go to WASM.
2. Keep running both the services codejail/wasm during the transition which could be an overhead for some people.

Reply:
> What could a rough estimated timeline for discontinuing the codejail if these experiments go smoothly with WASM. What provoked this thought was that during the transition from codejail to wasm we will have to:

It really depends on library support. I don't have a good sense of when it will land at the moment.

I think that when we're able to build something codejail-compatible in terms of library support, we would use our toggles to send problem custom response code to it by default–i.e. without having to alter the content.

So in other words, a phased rollout that looks like:

1. **Experiment Phase**
   Enable Wasm support for a simpler runtime image that doesn't have full compatibility with codejail. Authors can voluntarily opt into running on it by adding an attribute to their script tag. Scripts without the new attribute tag continue to go to codejail like they did before. (Lots of testing/discovery/iteration after this.)
2. **Migration Phase 1: Opt-in to Wasm, Default to codejail**
    Create Wasm runtime that is codejail compatible (once SciPy and other libs all build in `wasm32-wasi`). Use a CourseWaffleFlag to opt a few brave courses into the experiment. Fix bugs, possibly iterate on operational/performance issues.
3. **Migration Phase 2: Default to Wasm, Allow codejail**
    When we feel that things are solid, we can default to the codejail-compatible Wasm runtime for existing problems, but still allow courses to use codejail instead. That's historically how we've caught the long tail of weird edge case regressions.
4. **Remove codejail**
    At the end of this, we can turn codejail off entirely.

How exactly this works across our releases is something we'd still have to work out. I'd like to see Redwood release with experimental-but-still useful Wasm support. I don't think we'd see a viable migration path away from codejail until at least the S release (Sumac?). We'd definitely have at least one release (maybe two) where both are concurrently supported.

Reply:
Thanks for the details on this one. It looks like a reasonable plan.